src/node/orb.yml

```diff
@@ -122,11 +122,11 @@ commands:
       dir:
         type: string
         description: Filepath to your Node modules.
-        default: ${CIRCLE_WORKING_DIRECTORY}/node_modules
+        default: ~/project/node_modules
       cache-key:
         type: string
         description: File to use as a Node cache checksum.
-        default: "package.json"
+        default: "package-lock.json"
```

Comment:
This will only work if the project is using an `npm` lockfile. How will this work for Yarn-based projects? Can we detect the package manager in use and specify it's lockfile?

Reply:
Maybe a top level boolean allowing someone to say that they're using `yarn` as package-manager and using different behavior in that case?

Reply:
Yup. Cache strategy included would be great and it's always better to have good defaults. So if users specifically want to wipe the cache they still will be able to change the version.

> Another idea is to export a command

Was thinking about that exactly, because the other ways are more complicated than that.

Reply:
Or just eliminate the lockfile part of the cache string and assume the package manager will do the right thing. Then I could just pass in `yarn install` as its own step to the `steps` parameter of the `with-cache` command.

Reply:
Probably adding `use-loose-cache` param makes more sense for such cases. It will be the same as `use-strict-cache` without the checksum-ing.

```yaml
  steps:
    - when:
      condition: << parameters.use-loose-cache >>
      steps:
        - restore_cache:
          keys:
            - node-deps-<< parameters.cache-version >>-{{ .Branch }}

```

Reply:
@tunnckoCore if `yarn` already does the right thing, then why, if detect that it's `yarn`, and the lock file has changed since the previous run of the CI, do we need to wipe the cache?

Reply:
To provide a reason for using the lockfile checksum as part of the cache key, and, thereby, answer my own question:

In short, adding a checksum for a lockfile will, in most cases, improve the install performance of the package manager in cases where you find yourself executing a re-build for old commits on the same branch.

Assuming you have two commits, commit `A`, and commit `B`, and they each modify the lockfile.

Assuming you leave off the checksum of your project's lockfile.

You run a build for commit `A`. The contents of `node_modules` are cached.

You run a build for commit `B`. The cache from the build of `A` is restored, and your package manager spends time greater than 1 second mutating `node_modules` to match expectations.

You run a build for commit `A`. The cache from the build of `B` is restored, and your package manager spends time greater than 1 second mutating `node_modules` to match expectations.

If you _had_ used a checksum, and you relied on the fallback behavior of the restore cache step, then the build of `B`, would have fallen back to using the cache from `A`, because it had never stored a cache of it's own dependencies, and it would have taken the package manager more than 1 second to mutate the contents of the `node_modules` directory.

However, the second build of `A` would have re-used the cache from the first build of `A`, and the package manager would have taken _less_ than 1 second, because no mutation would have been necessary.

Therefore, using a checksum avoids the mutation when re-building old commits on the same branch.